SignalServiceKit/src/Contacts/OWSContactDiscoveryOperation.swift

```diff
@@ -0,0 +1,250 @@
+//
+//  Copyright (c) 2018 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+
+@objc(OWSContactDiscoveryOperation)
+class ContactDiscoveryOperation: OWSOperation {
+
+    let batchSize = 2048
+    let recipientIdsToLookup: [String]
+
+    @objc
+    var registeredRecipientIds: Set<String>
+
+    @objc
+    required init(recipientIdsToLookup: [String]) {
+        self.recipientIdsToLookup = recipientIdsToLookup
+        self.registeredRecipientIds = Set()
+
+        super.init()
+
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+        for batchIds in recipientIdsToLookup.chunked(by: batchSize) {
+            let batchOperation = LegacyContactDiscoveryBatchOperation(recipientIdsToLookup: batchIds)
+            self.addDependency(batchOperation)
+        }
+    }
+
+    // MARK: Mandatory overrides
+
+    // Called every retry, this is where the bulk of the operation's work should go.
+    override func run() {
+        Logger.debug("\(logTag) in \(#function)")
+
+        for dependency in self.dependencies {
+            guard let batchOperation = dependency as? LegacyContactDiscoveryBatchOperation else {
+                owsFail("\(self.logTag) in \(#function) unexpected dependency: \(dependency)")
+                continue
+            }
+
+            self.registeredRecipientIds.formUnion(batchOperation.registeredRecipientIds)
+        }
+
+        self.reportSuccess()
+    }
+}
+
+class LegacyContactDiscoveryBatchOperation: OWSOperation {
+
+    var registeredRecipientIds: Set<String>
+
+    private let recipientIdsToLookup: [String]
+    private var networkManager: TSNetworkManager {
+        return TSNetworkManager.shared()
+    }
+
+    // MARK: Initializers
+
+    required init(recipientIdsToLookup: [String]) {
+        self.recipientIdsToLookup = recipientIdsToLookup
+        self.registeredRecipientIds = Set()
+
+        super.init()
+
+        Logger.debug("\(logTag) in \(#function) with recipientIdsToLookup: \(recipientIdsToLookup.count)")
+    }
+
+    // MARK: OWSOperation Overrides
+
+    // Called every retry, this is where the bulk of the operation's work should go.
+    override func run() {
+        Logger.debug("\(logTag) in \(#function)")
+
+        var phoneNumbersByHashes: [String: String] = [:]
+
+        for recipientId in recipientIdsToLookup {
+            let hash = Cryptography.truncatedSHA1Base64EncodedWithoutPadding(recipientId)
+            phoneNumbersByHashes[hash] = recipientId
```

Comment:
assert that there aren't any collisions?

Reply:
done!